File: backend/app/api/v1/admin_applications.py
"""行政端:線上申請管理(幹部證明 `acert` / 郵局帳戶異動 `apostal`,各一把鍵)。

2026-07-17 需求方拍板的最小審核:狀態=審核中 → 處理中 → 已完成;
**只能往前,可跳過處理中**(D-25)、無退回;audit + notify 社團。
"""

from typing import Annotated

import sqlalchemy as sa
from fastapi import APIRouter, BackgroundTasks, Depends, Query, Request

from app.api.pagination import Pagination
from app.core.deps import CurrentUser, DbDep, client_ip, require_permission
from app.core.errors import conflict, not_found
from app.models import Club, OfficerCertificate, PostalAccountChange
from app.models.enums import ApplicationStatus
from app.schemas.activities import FileOut
from app.schemas.admin import AdminOfficerCertOut, AdminPostalChangeOut, ApplicationStatusIn
from app.schemas.common import ApiResponse
from app.services import audit, notify
from app.services import files as file_service

router = APIRouter(prefix="/admin", tags=["admin"])

# 證明與郵局各自一頁一把鍵(decisions.md D-11)
CertAdmin = Annotated[CurrentUser, Depends(require_permission("acert"))]
PostalAdmin = Annotated[CurrentUser, Depends(require_permission("apostal"))]

# 狀態機:審核中 → 處理中 → 已完成,**只能往前,可跳過**(D-25)——
# 承辦當場開完證明就直接結案,逼他先點一次「處理中」只是多一次點擊。不可回退。
_ALLOWED_NEXT = {
    ApplicationStatus.PENDING: (ApplicationStatus.PROCESSING, ApplicationStatus.COMPLETED),
    ApplicationStatus.PROCESSING: (ApplicationStatus.COMPLETED,),
}

# 幹部證明另有「已駁回」這個終態(D-37):不符資格的申請要有結束的方法,
# 否則只能一直掛在待處理裡。**郵局帳戶異動不跟進**,那邊沿用 _ALLOWED_NEXT。
_CERT_NEXT = {
    src: (*nxt, ApplicationStatus.DECLINED) for src, nxt in _ALLOWED_NEXT.items()
}

_STATUS_LABELS = {
    ApplicationStatus.PENDING: "審核中",
    ApplicationStatus.PROCESSING: "處理中",
    ApplicationStatus.COMPLETED: "已完成",
    ApplicationStatus.DECLINED: "已駁回",
}


def _status_order(model) -> sa.Case:
    return sa.case(
        (model.status == ApplicationStatus.PENDING, 0),
        (model.status == ApplicationStatus.PROCESSING, 1),
        else_=2,
    )


@router.get("/officer-certificates")
async def list_officer_certs(
    user: CertAdmin,
    db: DbDep,
    page: Pagination,
    status: ApplicationStatus | None = None,
    club_id: int | None = Query(None),
) -> ApiResponse[list[AdminOfficerCertOut]]:
    query = sa.select(OfficerCertificate, Club.name).join(
        Club, OfficerCertificate.club_id == Club.id
    )
    if status:
        query = query.where(OfficerCertificate.status == status)
    if club_id:
        query = query.where(OfficerCertificate.club_id == club_id)
    # 預設:審核中 → 處理中 → 完成,各組內申請日升冪(排序/篩選由前端做,不開 sort 參數)
    query = query.order_by(
        _status_order(OfficerCertificate), OfficerCertificate.created_at.asc(),
        OfficerCertificate.id,
    )

    total = await db.scalar(sa.select(sa.func.count()).select_from(query.subquery()))
    rows = await db.execute(query.offset(page.offset).limit(page.page_size))
    data = []
    for row, club_name in rows:
        out = AdminOfficerCertOut.model_validate(row)
        out.club_name = club_name
        data.append(out)
    return ApiResponse(data=data, meta=page.meta(total or 0))


@router.get("/postal-changes")
async def list_postal_changes(
    user: PostalAdmin,
    db: DbDep,
    page: Pagination,
    status: ApplicationStatus | None = None,
    club_id: int | None = Query(None),
) -> ApiResponse[list[AdminPostalChangeOut]]:
    query = sa.select(PostalAccountChange, Club.name).join(
        Club, PostalAccountChange.club_id == Club.id
    )
    if status:
        query = query.where(PostalAccountChange.status == status)
    if club_id:
        query = query.where(PostalAccountChange.club_id == club_id)
    query = query.order_by(
        _status_order(PostalAccountChange), PostalAccountChange.created_at.asc(),
        PostalAccountChange.id,
    )

    total = await db.scalar(sa.select(sa.func.count()).select_from(query.subquery()))
    rows = (await db.execute(query.offset(page.offset).limit(page.page_size))).all()
    # 存簿影本是核對局號帳號的依據,和單子一起回(整頁一次查,不逐列)
    passbooks = await file_service.files_by_subject(db, "postal_change", [r.id for r, _ in rows])
    data = []
    for row, club_name in rows:
        out = AdminPostalChangeOut.model_validate(row)
        out.club_name = club_name
        out.passbook = [FileOut.model_validate(f) for f in passbooks.get(row.id, [])]
        data.append(out)
    return ApiResponse(data=data, meta=page.meta(total or 0))


async def _advance_status(db, row, body: ApplicationStatusIn, allowed=_ALLOWED_NEXT):
    """只准往前的檢核(呼叫端先 with_for_update 取列)。

    與 `admin_maintenance` 的單步前進不同:那邊的「處理中」代表師傅真的在修,
    這裡的處理中只是承辦的工作註記,跳過它不會漏掉任何一次真實動作。
    """
    if body.status not in allowed.get(row.status, ()):
        raise conflict(
            f"無法從「{_STATUS_LABELS[row.status]}」變更為「{_STATUS_LABELS[body.status]}」",
            code="INVALID_STATUS_TRANSITION",
        )
    row.status = body.status


@router.post("/officer-certificates/{cert_id}/status")
async def update_officer_cert_status(
    cert_id: int,
    body: ApplicationStatusIn,
    user: CertAdmin,
    db: DbDep,
    request: Request,
    background: BackgroundTasks,
) -> ApiResponse[AdminOfficerCertOut]:
    row = await db.scalar(
        sa.select(OfficerCertificate).where(OfficerCertificate.id == cert_id).with_for_update()
    )
    if row is None:
        raise not_found("找不到幹部證明申請")
    await _advance_status(db, row, body, _CERT_NEXT)
    audit.record(
        db,
        action="officer_cert_status_updated",
        user=user,
        detail=f"officer_cert={row.id};status={body.status.value}",
        ip=client_ip(request),
    )
    await db.commit()

    club = await db.get(Club, row.club_id)
    # 駁回不附原因(承辦線下說明),通知只讓社團知道這張單結束了
    kind, title = {
        ApplicationStatus.COMPLETED: ("approve", "幹部證明已製作完成，請洽學務處領取"),
        ApplicationStatus.DECLINED: ("reject", "幹部證明已被駁回"),
    }.get(body.status, ("alert", "幹部證明處理中"))
    background.add_task(
        notify.club_event,
        kind,
        title,
        f"{club.name}:{row.term} {row.position.value} {row.applicant_name}",
        club.discord_webhook_url,
    )
    out = AdminOfficerCertOut.model_validate(row)
    out.club_name = club.name
    return ApiResponse(data=out)


@router.post("/postal-changes/{change_id}/status")
async def update_postal_change_status(
    change_id: int,
    body: ApplicationStatusIn,
    user: PostalAdmin,
    db: DbDep,
    request: Request,
    background: BackgroundTasks,
) -> ApiResponse[AdminPostalChangeOut]:
    row = await db.scalar(
        sa.select(PostalAccountChange)
        .where(PostalAccountChange.id == change_id)
        .with_for_update()
    )
    if row is None:
        raise not_found("找不到郵局帳戶異動申請")
    await _advance_status(db, row, body)
    audit.record(
        db,
        action="postal_change_status_updated",
        user=user,
        detail=f"postal_change={row.id};status={body.status.value}",
        ip=client_ip(request),
    )
    await db.commit()

    club = await db.get(Club, row.club_id)
    done = body.status == ApplicationStatus.COMPLETED
    background.add_task(
        notify.club_event,
        "approve" if done else "alert",
        "郵局帳戶異動已完成,請洽學務處" if done else "郵局帳戶異動處理中",
        f"{club.name}:{'、'.join(row.reasons)}",
        club.discord_webhook_url,
    )
    out = AdminPostalChangeOut.model_validate(row)
    out.club_name = club.name
    return ApiResponse(data=out)

File: frontend/src/api/adminApplications.ts
// 行政端線上申請管理 API 層(權限鍵 aapply):幹部證明/郵局帳戶異動。
// 兩張表各自伺服器端分頁(排序由後端固定);狀態流轉走後端狀態機:
// 僅允許單步前進(審核中→處理中→已完成),UI 也只開放下一步選項。
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import dayjs from 'dayjs'
import { api, apiPaged, qs } from './client'
import { useInvalidateBadges } from './badges'

export type ApplicationStatus = 'pending' | 'processing' | 'completed'

/** 單步前進狀態機(與後端 _NEXT_STATUS 對齊) */
export const NEXT_STATUS: Partial<Record<ApplicationStatus, ApplicationStatus>> = {
  pending: 'processing',
  processing: 'completed',
}

export interface OfficerCertRow {
  id: number
  club: string
  term: string
  position: string
  applicant: string
  date: string // 申請日 YYYY/MM/DD
  status: ApplicationStatus
}

export interface PostalChangeRow {
  id: number
  club: string
  reasons: string[]
  /** 事由以外全部選填(decisions.md D-07) */
  accountName?: string
  accountNumber?: string
  newAgentName?: string
  newAgentPhone?: string
  date: string
  status: ApplicationStatus
  /** 存簿影本(未歸檔者;下載走 GET /files/{id}) */
  passbook: { id: string; name: string }[]
}

interface AdminOfficerCertOut {
  id: number
  club_name: string
  term: string
  position: string
  applicant_name: string
  status: ApplicationStatus
  created_at: string
}

interface AdminPostalChangeOut {
  id: number
  club_name: string
  reasons: string[]
  account_name: string | null
  account_number: string | null
  new_agent_name: string | null
  new_agent_phone: string | null
  status: ApplicationStatus
  created_at: string
  passbook: { id: string; original_name: string }[]
}

const toCertRow = (c: AdminOfficerCertOut): OfficerCertRow => ({
  id: c.id,
  club: c.club_name,
  term: c.term,
  position: c.position,
  applicant: c.applicant_name,
  date: dayjs(c.created_at).format('YYYY/MM/DD'),
  status: c.status,
})

const toPostalRow = (p: AdminPostalChangeOut): PostalChangeRow => ({
  id: p.id,
  club: p.club_name,
  reasons: p.reasons,
  accountName: p.account_name ?? undefined,
  accountNumber: p.account_number ?? undefined,
  newAgentName: p.new_agent_name ?? undefined,
  newAgentPhone: p.new_agent_phone ?? undefined,
  date: dayjs(p.created_at).format('YYYY/MM/DD'),
  status: p.status,
  passbook: (p.passbook ?? []).map((f) => ({ id: f.id, name: f.original_name })),
})

const keys = {
  all: ['adminApplications'] as const,
  certs: (page: number) => ['adminApplications', 'certs', page] as const,
  postal: (page: number) => ['adminApplications', 'postal', page] as const,
  pendingTotal: (kind: ApplicationKind) => ['adminApplications', 'pendingTotal', kind] as const,
}

export const APPLICATIONS_PAGE_SIZE = 50

export type ApplicationKind = 'cert' | 'postal'

const KIND_PATH: Record<ApplicationKind, string> = {
  cert: '/admin/officer-certificates',
  postal: '/admin/postal-changes',
}

const pagePath = (kind: ApplicationKind, page: number) =>
  `${KIND_PATH[kind]}${qs({ page, page_size: APPLICATIONS_PAGE_SIZE })}`

/** 兩張表皆為後端固定排序(審核中→處理中→完成,組內申請日升冪),無 sort 參數 */
export function useAdminOfficerCerts(page: number) {
  return useQuery({
    queryKey: keys.certs(page),
    queryFn: () =>
      apiPaged<AdminOfficerCertOut[]>(pagePath('cert', page)).then(({ data, total }) => ({
        rows: data.map(toCertRow),
        total,
      })),
  })
}

export function useAdminPostalChanges(page: number) {
  return useQuery({
    queryKey: keys.postal(page),
    queryFn: () =>
      apiPaged<AdminPostalChangeOut[]>(pagePath('postal', page)).then(({ data, total }) => ({
        rows: data.map(toPostalRow),
        total,
      })),
  })
}

/** 單類待處理件數(page_size=1 只取 meta.total;分頁後算不出全域數字)。
 *  拆頁之後兩頁各算各的 —— 合計會讓只有一把鍵的承辦看到自己讀不到的那類件數 */
function usePendingTotal(kind: ApplicationKind) {
  return useQuery({
    queryKey: keys.pendingTotal(kind),
    queryFn: async () =>
      (await apiPaged<unknown[]>(`${KIND_PATH[kind]}${qs({ status: 'pending', page_size: 1 })}`))
        .total,
  })
}

export const usePendingCertTotal = () => usePendingTotal('cert')
export const usePendingPostalTotal = () => usePendingTotal('postal')

export function useApplicationStatusMutation() {
  const qc = useQueryClient()
  const invalidateBadges = useInvalidateBadges()
  return useMutation({
    mutationFn: ({ kind, id, status }: { kind: ApplicationKind; id: number; status: ApplicationStatus }) =>
      api(`${KIND_PATH[kind]}/${id}/status`, {
        method: 'POST',
        body: JSON.stringify({ status }),
      }),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: keys.all })
      invalidateBadges()
    },
  })
}
